### PR TITLE
Allow user to specify the address to bind to

### LINKED ISSUE #2424: Allow starting materialized by binding to specified network interface instead of always 0.0.0.0
When running materialized locally, I'd like to be able to bind to a specific network interface and not just default to `0.0.0.0`, such that the following would work:

```
$ materialized --data-directory /tmp/mzdata --addr 127.0.0.1:6875
=======================================================================
Thank you for trying Materialize!

We are interested in any and all feedback you have, which may be able
to improve both our software and your queries! Please reach out at:

    Web: https://materialize.io
    GitHub issues: https://github.com/MaterializeInc/materialize/issues
    Email: support@materialize.io
    Twitter: @MaterializeInc
=======================================================================

materialized v0.1.3 (ac3543646) listening on 127.0.0.1:6875...
```

Instead, `--addr` doesn't exist, and we get:

```
materialized v0.1.3 (ac3543646) listening on 0.0.0.0:6875...
```

One can get around this by putting it in a docker container, but that, of course, has it's own problems for local development.

### PR DESCRIPTION
**Pull request by umanwizard**
Fixes #2424

### CODE REVIEW

File: src/materialized/bin/materialized.rs [hunk 1]
```diff
@@ -33,6 +33,7 @@ use failure::{bail, format_err, ResultExt};
 use lazy_static::lazy_static;
 use log::trace;
 use once_cell::sync::OnceCell;
+use std::str::FromStr;
```
Comment (benesch) [inline src/materialized/bin/materialized.rs]: Please move this to the `std` block above.

File: src/materialized/bin/materialized.rs [hunk 2]
```diff
@@ -112,6 +113,12 @@ fn run() -> Result<(), failure::Error> {
         "where materialized will write logs (default <data directory>/materialized.log)",
         "PATH",
     );
+    opts.optopt(
+        "",
+        "bind-to",
```
Comment (benesch) [inline src/materialized/bin/materialized.rs]: Can we call this `listen-addr`? That's what I typically see used as the name and it parallels the PostgreSQL connection parameter: https://www.postgresql.org/docs/current/runtime-config-connection.html

File: src/materialized/bin/materialized.rs [hunk 3]
```diff
@@ -158,6 +165,10 @@ fn run() -> Result<(), failure::Error> {
     let processes = popts.opt_get_default("processes", 1)?;
     let address_file = popts.opt_str("address-file");
     let gather_metrics = !popts.opt_present("no-prometheus");
+    let bind_to = popts
+        .opt_str("bind-to")
+        .map(|s| SocketAddr::from_str(&s))
+        .transpose()?;
```
Comment (benesch) [inline src/materialized/bin/materialized.rs]: This'll be cleaner with https://docs.rs/getopts/0.2.21/getopts/struct.Matches.html#method.opt_get.

File: src/materialized/lib.rs [hunk 4]
```diff
@@ -103,6 +103,9 @@ pub struct Config {
     /// Whether to collect metrics. If enabled, metrics can be collected by
     /// e.g. Prometheus via the `/metrics` HTTP endpoint.
     pub gather_metrics: bool,
+    /// The IP address and port to listen on -- defaults to 0.0.0.0:<addr_port>,
+    /// where <addr_port> is the address of this process's entry in `addresses`.
+    pub bind_to: Option<SocketAddr>,
```
Comment (benesch) [inline src/materialized/lib.rs]: And then this can be `listen_addr`...

File: src/materialized/lib.rs [hunk 5]
```diff
@@ -177,13 +180,15 @@ pub fn serve(mut config: Config) -> Result<Server, failure::Error> {
...
 
     // Initialize network listener.
-    let listen_addr = SocketAddr::new(
-        match config.addresses[config.process].ip() {
-            IpAddr::V4(_) => IpAddr::V4(Ipv4Addr::UNSPECIFIED),
-            IpAddr::V6(_) => IpAddr::V6(Ipv6Addr::UNSPECIFIED),
-        },
-        config.addresses[config.process].port(),
-    );
+    let listen_addr = config.bind_to.unwrap_or_else(|| {
```
Comment (benesch) [inline src/materialized/lib.rs]: ...which even matches what we call it internally already.

File: src/materialized/bin/materialized.rs (hunk 1 above)
Comment (umanwizard) [reply src/materialized/bin/materialized.rs]: Actually, it's not needed at all if I switch to using `opt_get` as you suggested.

**Comment by cuongdo** [comment]
@umanwizard could you add a relevant release note to `doc/user/release-notes.md`?

**Comment by umanwizard** [comment]
@cuongdo thanks for the reminder, #2489

**Pull request merged**